### PR TITLE
feat(tekton): view pipelinerun visualization on click of pipelinerun

### PR DESCRIPTION
**Pull request by debsmita1**
**Resolves:**
https://github.com/janus-idp/backstage-plugins/issues/408

**Screenshots:**
Screenshot 2023-07-05 at 2 37 17 PM
Screenshot 2023-07-05 at 2 37 27 PM

https://github.com/janus-idp/backstage-plugins/assets/22490998/54afb29c-c404-4729-ba7e-09c92b1ed3ef

### CODE REVIEW

File: plugins/tekton/src/plugin.ts [hunk 1]
```diff
@@ -4,34 +4,37 @@ import '@patternfly/patternfly/patternfly-charts-theme-dark.css';
...
-    component: () => import('./components/Tekton').then(m => m.TektonComponent),
+    component: () => import('./components/Tekton/Router').then(m => m.Router),
     mountPoint: rootRouteRef,
   }),
 );
 
 export const LatestPipelineRun = tektonPlugin.provide(
-  createRoutableExtension({
+  createComponentExtension({
     name: 'LatestPipelineRun',
```
Comment (invincibleJai) [inline plugins/tekton/src/plugin.ts:32]: should we rename this as well to `PipelineRun`

File: plugins/tekton/src/utils/pipelineRun-utils.ts [hunk 2]
```diff
@@ -97,6 +97,16 @@ export const getLatestPipelineRun = (
   return null;
 };
 
+export const getPipelineRun = (
```
Comment (invincibleJai) [inline plugins/tekton/src/utils/pipelineRun-utils.ts:100]: can we have unit tests for this?

File: plugins/tekton/src/components/pipeline-topology/PipelineVisualization.tsx [hunk 3]
```diff
@@ -147,12 +192,31 @@ export const PipelineVisualization = ({
         </div>
       ) : (
         <div
-          data-testid="pipeline-visualization"
+          data-testid={
```
Comment (invincibleJai) [inline plugins/tekton/src/components/pipeline-topology/PipelineVisualization.tsx:195]: can we avoid this? the component should be declarative and ideally not good to change test id conditionally

File: plugins/tekton/src/components/pipeline-topology/PipelineVisualization.tsx [hunk 4]
```diff
@@ -147,12 +192,31 @@ export const PipelineVisualization = ({
...
+          }
           className="bs-tkn-pipeline-visualization__layout"
         >
           <PipelineLayout model={model} />
         </div>
       )}
     </WrapperInfoCard>
   );
+
+  return pipelineRun ? (
```
Comment (invincibleJai) [inline plugins/tekton/src/components/pipeline-topology/PipelineVisualization.tsx:208]: this component has grown, can we refactor it?

File: plugins/tekton/src/components/pipeline-topology/PipelineVisualization.test.tsx [hunk 5]
```diff
@@ -42,7 +48,40 @@ describe('PipelineVisualization', () => {
...
+      </TektonResourcesContext.Provider>,
+    );
+    expect(
+      screen.getByText(mockKubernetesPlrResponse.pipelineruns[1].metadata.name),
+    ).toBeInTheDocument();
+    expect(screen.getByText('Pipeline Run')).toBeInTheDocument();
+
+    expect(screen.getByText('Back to PipelineRun list')).toBeInTheDocument();
+
+    expect(queryByTestId('pipelineRun-visualization')).not.toBeNull();
```
Comment (invincibleJai) [inline plugins/tekton/src/components/pipeline-topology/PipelineVisualization.test.tsx:84]: can be consistent with the above assertions in same it block

```suggestion
    expect(queryByTestId('pipelineRun-visualization')).toBeInTheDocument();;
```

File: plugins/tekton/src/components/pipeline-topology/PipelineVisualization.test.tsx [hunk 6]
```diff
@@ -42,7 +48,40 @@ describe('PipelineVisualization', () => {
       screen.getByText(mockKubernetesPlrResponse.pipelineruns[1].metadata.name),
     ).toBeInTheDocument();
 
-    expect(queryByTestId('pipeline-visualization')).not.toBeNull();
+    expect(queryByTestId('latest-pipelineRun-visualization')).not.toBeNull();
+  });
+
+  it('should render the pipeline run visualization when pipelineRun name exists in the params', async () => {
```
Comment (invincibleJai) [inline plugins/tekton/src/components/pipeline-topology/PipelineVisualization.test.tsx:54]: this doesn't need to be async

```suggestion
  it('should render the pipeline run visualization when pipelineRun name exists in the params', () => {
```

File: plugins/tekton/src/components/pipeline-topology/PipelineVisualization.test.tsx [hunk 7]
```diff
@@ -16,8 +17,13 @@ jest.mock('@backstage/plugin-catalog-react', () => ({
...
 }));
 
+jest.mock('@backstage/core-plugin-api', () => ({
+  ...jest.requireActual('@backstage/core-plugin-api'),
+  useRouteRef: () => () => '/xyz',
+}));
+
 describe('PipelineVisualization', () => {
-  it('should render the pipeline run visualization when pipelineRun exists', async () => {
+  it('should render the latest pipeline run visualization when pipelineRun exists', async () => {
```
Comment (invincibleJai) [inline plugins/tekton/src/components/pipeline-topology/PipelineVisualization.test.tsx:26]: component should be declarative, and should just render PLR without the notion of this being the latest or old one.

File: plugins/tekton/src/components/PipelineRunList/PipelineRunHeader.tsx [hunk 8]
```diff
@@ -20,11 +20,13 @@ export const PipelineRunHeader: TableColumn[] = [
     field: 'metadata.name',
     type: 'string',
     render: (row): React.ReactNode => (
-      <ResourceBadge
-        color={tektonGroupColor.value}
-        abbr="PLR"
-        name={(row as PipelineRunKind)?.metadata?.name || ''}
-      />
+      <Link to={`pipelineRun/${(row as PipelineRunKind)?.metadata?.name}`}>
```
Comment (invincibleJai) [inline plugins/tekton/src/components/PipelineRunList/PipelineRunHeader.tsx:23]: `(row as PipelineRunKind)?.metadata?.name` id duplicated here, good to keep it in const

File: plugins/tekton/src/plugin.ts (hunk 1 above)
Comment (debsmita1) [reply plugins/tekton/src/plugin.ts:32]: The LatestPipelineRun & PipelineRun are rendered on 2 different routes

File: plugins/tekton/src/components/PipelineRunList/PipelineRunList.tsx [hunk 9]
```diff
@@ -64,7 +67,16 @@ const PipelineRunList = () => {
...
+        <Table
+          title="Pipeline Runs"
+          data={[]}
+          columns={PipelineRunHeader}
+          emptyContent={
+            <div data-testid="no-pipeline-runs" className={classes.empty}>
+              No Pipeline Runs found
+            </div>
+          }
+        />
```
Comment (invincibleJai) [inline plugins/tekton/src/components/PipelineRunList/PipelineRunList.tsx:79]: do we need this or can we get rid of this if block itself?

File: plugins/tekton/src/components/Tekton/index.ts [hunk 10]
```diff
@@ -1 +0,0 @@
-export { TektonComponent } from './TektonComponent';
```
Comment (invincibleJai) [inline plugins/tekton/src/components/Tekton/index.ts:1]: we don't need this?

File: plugins/tekton/src/components/pipeline-topology/PipelineVisualizationCard.test.tsx [hunk 11]
```diff
@@ -16,8 +17,41 @@ jest.mock('@backstage/plugin-catalog-react', () => ({
...
+      clusters: [],
+      setSelectedCluster: () => {},
+    };
+    const { queryByTestId } = render(
+      <TektonResourcesContext.Provider value={mockContextData}>
+        <PipelineVisualizationCard linkTekton={false} />
+      </TektonResourcesContext.Provider>,
+    );
+    expect(
+      screen.getByText(mockKubernetesPlrResponse.pipelineruns[1].metadata.name),
```
Comment (invincibleJai) [inline plugins/tekton/src/components/pipeline-topology/PipelineVisualizationCard.test.tsx:48]: we can use `getByText` from the render 

`const { getByText } = render( ... )`

File: plugins/tekton/src/components/pipeline-topology/PipelineVisualizationCard.test.tsx [hunk 12]
```diff
@@ -16,8 +17,41 @@ jest.mock('@backstage/plugin-catalog-react', () => ({
...
+    ).toBeInTheDocument();
+
+    expect(queryByTestId('latest-pipelineRun-visualization')).not.toBeNull();
```
Comment (invincibleJai) [inline plugins/tekton/src/components/pipeline-topology/PipelineVisualizationCard.test.tsx:51]: nit: 

```suggestion
    expect(queryByTestId('latest-pipelineRun-visualization')).toBeInTheDocument();
```

File: plugins/tekton/src/plugin.ts [hunk 13]
```diff
@@ -4,34 +4,37 @@ import '@patternfly/patternfly/patternfly-charts-theme-dark.css';
...
     name: 'TektonPage',
-    component: () => import('./components/Tekton').then(m => m.TektonComponent),
+    component: () => import('./components/Tekton/Router').then(m => m.Router),
     mountPoint: rootRouteRef,
   }),
 );
 
 export const LatestPipelineRun = tektonPlugin.provide(
-  createRoutableExtension({
+  createComponentExtension({
```
Comment (invincibleJai) [inline plugins/tekton/src/plugin.ts:31]: curious what's the diff between `createRoutableExtension` and `createComponentExtension`?

**Comment by invincibleJai** [comment]
in the case of multiple clusters with the first cluster having no PLR and the second having few and I select PLR from the list view on cluster 2 then I see empty PLR viz as context is missing.

**Comment by debsmita1** [comment]
> in the case of multiple clusters with the first cluster having no PLR and the second having few and I select PLR from the list view on cluster 2 then I see empty PLR viz as context is missing.

Fixed it ! Added Gif in the PR description

File: plugins/tekton/src/routes.ts [hunk 14]
```diff
@@ -1,5 +1,11 @@
...
+import { createRouteRef, createSubRouteRef } from '@backstage/core-plugin-api';
 
 export const rootRouteRef = createRouteRef({
   id: 'tekton',
 });
+
+export const pipelineRunRouteRef = createSubRouteRef({
+  id: 'pipeline-run-visualization',
+  parent: rootRouteRef,
+  path: '/cluster/:clusterName/pipelineRun/:pipelineRunName',
```
Comment (invincibleJai) [inline plugins/tekton/src/routes.ts:10]: here it says `:clusterName` but what's passed in number i.e index https://github.com/janus-idp/backstage-plugins/pull/506/files#diff-cd65d6b98bf0c3daff8d2edaaad8e013f7a55c35d6df69374fca9a4c571c710eR28

Comment (debsmita1) [reply plugins/tekton/src/routes.ts:10]: `clusters[selectedCluster as number]` is the name of the selected cluster

Comment (invincibleJai) [reply plugins/tekton/src/routes.ts:10]: ah my bad, thanks 👍

**Comment by invincibleJai** [comment]
in dev mode am not able to click on PLR to see the PLR Viz 

image

**Review by invincibleJai — APPROVED** [review]
Thanks @debsmita1, this looks good.

**Pull request merged**